typing/rec_check.ml

```diff
@@ -18,145 +18,161 @@ open Types
-
-    let list_matching p t =
-      let r = ref [] in
-      M.iter (fun id v -> if p v then r := id :: !r) t;
-      !r
-
-    let unguarded =
-      list_matching (function Unguarded | Dereferenced -> true | _ -> false)
-
-    let dependent =
-      list_matching (function _ -> true)
-  end
-
-  module Env =
-  struct
-    (* A typing environment maps identifiers to types *)
-    type env = Use.t Ident.tbl
-
-    let empty = Ident.empty
-
-    let join x y =
-      let r =
-      Ident.fold_all
-        (fun id v tbl ->
-           let v' = try Ident.find_same id tbl with Not_found -> Use.empty in
-           Ident.add id (Use.join v v') tbl)
-        x
-        y
-      in
-      r
-  end
+module Mode = struct
+  (** For an expression in a program, its "usage mode" represents
+     static information about how the value produced by the expression
+     will be used by the context around it. *)
+  type t =
+    | Dereference
+    (** A [Dereference] context consumes, inspects and uses the value
+        in arbitrary ways. Such a value must be fully defined at the point
+        of usage, it cannot be defined mutually-recursively with its context. *)
+
+    | Return
+    (** A [Return] context returns its value without further inspection.
+        This value cannot be defined mutually-recursively with its context,
+        as there is a risk of self-loop: in [let rec x = y and y = x], the
+        two definitions use a single variable in [Return] context. *)
+
+    | Guard
+    (** A [Guard] context returns the value as a member of a data structure,
+        for example a variant constructor or record. The value can safely be
+        defined mutually-recursively with their context, for example in
+        [let rec li = 1 :: li].
+        When these subexpressions participate in a cyclic definition,
+        this definition is productive/guarded.
+
+        The [Guard] mode is also used when a value is not dereferenced,
+        it is returned by a sub-expression, but the result of this
+        sub-expression is discarded instead of being returned.
+        For example, the subterm [?] is in a [Guard] context
+        in [let _ = ? in e] and in [?; e].
+        When these subexpressions participate in a cyclic definition,
+        they cannot create a self-loop.
+    *)
+
+    | Delay
+    (** A [Delay] context can be fully evaluated without evaluting its argument,
+        which will only be needed at a later point of program execution. For
+        example, [fun x -> ?] or [lazy ?] are [Delay] contexts. *)
+
+    | Unused
+    (** [Unused] is for subexpressions that are not used at all during
+       the evaluation of the whole program. This is the mode of
+       a variable in an expression in which it does not occur. *)
+
+  (* Returns the most conservative mode of the two arguments.
+
+     Dereference < Return < Guard < Delay < Unused
+
+     In judgments we write (m + m') for (join m m').
+  *)
+  let join m m' =
```

Comment:
This is just `Pervasives.min`, and I think it would be fine to define it that way (with a comment).

Reply:
Meh, personally I dislike making use of the barely-defined variant ordering in this way, so I will keep the current definition.

Reply:
Those aren't the only alternatives.  Another possibility is to define a mapping to integers:

```ocaml
  let rank = function
    | Dereference -> 0
    | Return -> 1
    | Guard -> 2
    | Delay -> 3
    | Unused -> 4
```
and then use that to define `join` in the obvious way.

This is simpler than the current definition of `join`, not least because there's no ordering in the cases.  And it turns the mode ordering (which is currently only defined in a comment) into code.